docs/source/reference-core.rst

```diff
@@ -1245,58 +1245,320 @@ Broadcasting an event with :class:`Event`
+
+.. literalinclude:: reference-core/channels-fan-in-fan-out-broken.py
+
+The two producers, A and B, send 3 messages apiece. These are then
+randomly distributed between the two producers, X and Y. So we're
+hoping to see some output like:
+
+.. code-block:: none
+
+   consumer Y got value '0 from producer B'
+   consumer X got value '0 from producer A'
+   consumer Y got value '1 from producer A'
+   consumer Y got value '1 from producer B'
+   consumer X got value '2 from producer B'
+   consumer X got value '2 from producer A'
+
+However, on most runs, that's not what happens – the first part of the
+output is OK, and then when we get to the end the program crashes with
+:exc:`ClosedResourceError`. If you run the program a few times, you'll
+see that sometimes the traceback shows ``send`` crashing, and other
+times it shows ``receive`` crashing, and you might even find that on
+some runs it doesn't crash at all.
+
+Here's what's happening: suppose that producer A finishes first. It
+exits, and its ``async with`` block closes the ``send_channel``. But
+wait! Producer B was still using that ``send_channel``... so when it
+calls ``send``, it gets a :exc:`ClosedResourceError`.
+
+Sometimes, though if we're lucky, the two producers might finish at
+the same time (or close enough), so they both make their last ``send``
+before either of them closes the ``send_channel``.
+
+But, even if that happens, we're not out of the woods yet! After the
+producers exit, the two consumers race to be the first to notice that
+the ``send_channel`` has closed. Suppose that X wins the race. It
+exits its ``async for`` loop, then exits the ``async with`` block...
+and closes the ``receive_channel``, while Y is still using it. Again,
+this causes a crash.
+
+We could avoid this by using some complicated bookkeeping to make sure
+that only the *last* producer and the *last* consumer close their
+channel endpoints... but that would be tiresome and fragile.
+Fortunately, there's a better way! Here's a fixed version of our
+program above:
+
+.. literalinclude:: reference-core/channels-fan-in-fan-out-fixed.py
+   :emphasize-lines: 7, 9, 10, 12, 13
+
+What we're doing here is taking advantage of the
+:meth:`SendChannel.clone <trio.abc.SendChannel.clone>` and
+:meth:`ReceiveChannel.clone <trio.abc.ReceiveChannel.clone>` methods.
+What these do is create copies of our endpoints, that act just like
+the original – except that they can be closed independently. And the
+underlying channel is only closed after *all* the clones have been
+closed. So this completely solves our problem with shutdown, and if
+you run this program, you'll see it print its six lines of output and
+then exits cleanly.
+
+Notice a small trick we use: the code in ``main`` that sets up the
+channels in the first place uses an ``async with`` block to close the
+original objects after passing on clones to all the tasks. We could
+alternatively have passed on the original objects into the tasks,
```

Comment:
"into one of the tasks"? not sure, so will mark as resolved, since the broken example below clarifies what's different in the finicky example.

Reply:
Rewrote this section a bit